File: java/src/main/java/io/github/mb1lal/playwright/ui/pages/heroku/AddRemoveElementsPage.java
package io.github.mb1lal.playwright.ui.pages.herokuapp;

import io.github.mb1lal.playwright.base.TestContext;
import io.github.mb1lal.playwright.ui.pages.BasePage;
import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import lombok.extern.slf4j.Slf4j;

/**
 * Add/Remove Elements page ({@code /add_remove_elements/}).
 */
@Slf4j
public class AddRemoveElementsPage extends BasePage {

    private final Locator addButton;
    private final Locator deleteButtons;

    public AddRemoveElementsPage(Page page, TestContext testContext) {
        super(page, testContext);
        this.addButton = page.locator("button[onclick='addElement()']");
        this.deleteButtons = page.locator(".added-manually");
    }

    public void addElements(int times) {
        for (int i = 0; i < times; i++) {
            click(addButton, "Add Element button");
        }
    }

    public int deleteCount() {
        return getElementCount(deleteButtons, "Delete buttons");
    }

    public void assertDeleteVisible(boolean expected) {
        boolean visible = deleteCount() > 0
                && deleteButtons.first().isVisible();
        if (expected) {
            assertIsVisible(deleteButtons.first(), "Delete button");
        } else {
            io.github.mb1lal.playwright.utils.AssertionHelper.assertFalse(visible, "Delete button should be gone");
        }
    }

    public void deleteFirst() {
        click(deleteButtons.first(), "first Delete button");
    }
}

File: typescript/src/pages/heroku/add-remove-elements.page.ts
import { BasePage } from '../base.page';

/** Add/Remove Elements page (/add_remove_elements/). */
export class AddRemoveElementsPage extends BasePage {
  private readonly deleteButtons = this.page.locator('.added-manually');

  async addElements(times: number): Promise<void> {
    for (let i = 0; i < times; i++) {
      await this.page.locator("button[onclick='addElement()']").click();
    }
  }

  async deleteCount(): Promise<number> {
    return this.deleteButtons.count();
  }

  async isDeleteVisible(): Promise<boolean> {
    return (await this.deleteCount()) > 0 && (await this.deleteButtons.first().isVisible());
  }

  async deleteFirst(): Promise<void> {
    await this.deleteButtons.first().click();
  }
}
